sdk/provisioning/Azure.Provisioning.CloudMachine/src/Azd.cs

```diff
@@ -0,0 +1,141 @@
+        mainBicep.Add(location);
+        mainBicep.Add(principalId);
+        mainBicep.Build().Save(infraDirectory);
+
+        WriteMainParametersFile(infraDirectory);
+    }
+
+    private static void WriteMainParametersFile(string infraDirectory)
+    {
+        File.WriteAllText(Path.Combine(infraDirectory, $"{MainBicepName}.parameters.json"),
+        """
+        {
+          "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+          "contentVersion": "1.0.0.0",
+          "parameters": {
+            "environmentName": {
+              "value": "${AZURE_ENV_NAME}"
+            },
+            "location" : {
+              "value" : "${AZURE_LOCATION}"
+            },
+            "principalId": {
+              "value": "${AZURE_PRINCIPAL_ID}"
+            }
+          }
+        }
+        """);
+    }
+
+    internal static string ReadOrCreateCmid()
+    {
+        string appsettings = Path.Combine(".", "appsettings.json");
+
+        string? cmid;
+        if (!File.Exists(appsettings))
+        {
+            cmid = GenerateCloudMachineId();
+
+            using FileStream file = File.OpenWrite(appsettings);
+            Utf8JsonWriter writer = new Utf8JsonWriter(file);
+            writer.WriteStartObject();
+            writer.WritePropertyName("CloudMachine"u8);
+            writer.WriteStartObject();
+            writer.WriteString("ID"u8, cmid);
+            writer.WriteEndObject();
+            writer.WriteEndObject();
+            writer.Flush();
+            return cmid;
+        }
+
+        using FileStream json = File.OpenRead(appsettings);
+        JsonElement je = JsonDocument.Parse(json).RootElement;
+        // attempt to read CM configuration from existing configuration file
+        if (je.TryGetProperty("CloudMachine"u8, out JsonElement cm))
+        {
+            if (!cm.TryGetProperty("ID"u8, out JsonElement id))
+            {
+                throw new NotImplementedException();
+            }
+            cmid = id.GetString();
+            if (cmid == null)
+                throw new NotImplementedException();
+            return cmid;
+        }
+        else
+        { // add CM configuration to existing file
+            JsonNode? root = JsonValue.Parse(appsettings);
+            if (root is null)
+                throw new NotImplementedException();
+
+            if (root is not JsonObject obj)
+                throw new NotImplementedException();
+
+            var cmProperties = new JsonObject();
+            cmid = GenerateCloudMachineId();
+            cmProperties.Add("ID", cmid);
+            obj.Add("CloudMachine", cmProperties);
+        }
+
+        return cmid;
+
+        static string GenerateCloudMachineId()
+        {
+            var guid = Guid.NewGuid();
+            var guidString = guid.ToString("N");
+            var cnId = "cm" + guidString.Substring(0, 15); // we can increase it to 20, but the template name cannot be that long
```

Comment:
We've got automatic Name trimming if you want to make this super long and just let it truncate on the tiny name resources.

Reply:
I want to be explicit about these CM names. This is the algo with have been using since 1+ year ago